src/main/java/org/wildfly/security/password/spec/DefaultPasswordSpecEncoding.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * JBoss, Home of Professional Open Source
+ *
+ * Copyright 2015 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.password.spec;
+
+import org.wildfly.security._private.ElytronMessages;
+import org.wildfly.security.password.Password;
+import org.wildfly.security.password.PasswordFactory;
+import org.wildfly.security.password.interfaces.ClearPassword;
+import org.wildfly.security.password.interfaces.DigestPassword;
+import org.wildfly.security.util.ByteIterator;
+import org.wildfly.security.util.ByteStringBuilder;
+import org.wildfly.security.util.CodePointIterator;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+
+/**
+ * Provide methods for encoding and decoding of {@link PasswordSpec}.
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public final class DefaultPasswordSpecEncoding {
+
+    private static final int SALT_SIZE = 64;
+
+    // the value for each identifier must be kept and can not change
+    private static final byte CLEAR_PASSWORD_SPEC_ID = 1;
+    private static final byte DIGEST_PASSWORD_SPEC_ID = 2;
+    private static final byte HASH_PASSWORD_SPEC_ID = 3;
+    private static final byte SALTED_HASH_PASSWORD_SPEC_ID = 4;
+    private static final byte ITERATED_SALTED_HASH_SPEC_ID = 5;
+
+    private DefaultPasswordSpecEncoding() {}
+
+    /**
+     * Encode the given password to a byte array.
+     *
+     * @param password the password to encode
+     * @return a byte array representing the encoded password or null if no encoder was capable to encode the given password
+     */
+    public static byte[] encode(Password password) throws NoSuchAlgorithmException, InvalidKeySpecException {
```

Comment:
If this method is retained, then I suggest that instead of doing isInstance checks (which BTW are better written using ``instanceof``), you can do:

```
if (passwordFactory.convertibleToKeySpec(password, ClearPasswordSpec.class)) {
    return encodeClearPasswordSpec(passwordFactory.getKeySpec(ClearPasswordSpec.class));
} ...
```

This way you can support as-yet-unknown types.